File: etl/code scala/etl_loading.scala
import org.apache.spark.sql.{SparkSession, Window, SaveMode}
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._

object AdventureWorksETLCombined {
  def main(args: Array[String]): Unit = {
    // 1. Initialisation de la session Spark
    val spark = SparkSession.builder
      .appName("AdventureWorks ETL Combined")
      // Assurez-vous d'avoir les JARs pour SQL Server et PostgreSQL
      .config("spark.jars", "/app/lib/mssql-jdbc-12.8.1.jre11.jar,/app/lib/postgresql-42.7.3.jar")
      .getOrCreate()

    // 2. Paramètres JDBC pour SQL Server (Source)
    val jdbcUrlSqlServer = "jdbc:sqlserver://sqlserver:1433;databaseName=AdventureWorks;encrypt=true;trustServerCertificate=true"
    val connectionPropertiesSqlServer = new java.util.Properties()
    connectionPropertiesSqlServer.put("user", "SA")
    connectionPropertiesSqlServer.put("password", "password123?")
    connectionPropertiesSqlServer.put("driver", "com.microsoft.sqlserver.jdbc.SQLServerDriver")

    // 3. Paramètres JDBC pour PostgreSQL (Cible)
    // Utilisez 'postgresql:5432' si vous exécutez dans le même réseau Docker Compose
    // Utilisez 'host.docker.internal:5432' si vous exécutez Spark localement et PostgreSQL est dans Docker
    val jdbcUrlPg = "jdbc:postgresql://postgresql:5432/adventure_warehouse"
    // val jdbcUrlPg = "jdbc:postgresql://host.docker.internal:5432/adventure_warehouse" // Alternative pour local
    val connectionPropertiesPg = new java.util.Properties()
    connectionPropertiesPg.put("user", "postgres")
    connectionPropertiesPg.put("password", "pass")
    connectionPropertiesPg.put("driver", "org.postgresql.Driver")

    println("Starting full ETL pipeline from AdventureWorks to PostgreSQL...")

    // --- PHASE 1: Extraction des tables sources depuis SQL Server ---
    println("\n--- Phase 1: Extraction des données sources ---")

    println("Extracting Sales.Customer...")
    val dfCustomer = spark.read.jdbc(jdbcUrlSqlServer, "Sales.Customer", connectionPropertiesSqlServer)

    println("Extracting Person.Person...")
    val dfPerson = spark.read.jdbc(jdbcUrlSqlServer, "Person.Person", connectionPropertiesSqlServer)

    println("Extracting Person.PersonPhone...")
    val dfPersonPhone = spark.read.jdbc(jdbcUrlSqlServer, "Person.PersonPhone", connectionPropertiesSqlServer)

    println("Extracting Production.Product...")
    val dfProduct = spark.read.jdbc(jdbcUrlSqlServer, "Production.Product", connectionPropertiesSqlServer)

    println("Extracting Production.ProductCategory...")
    val dfProductCategory = spark.read.jdbc(jdbcUrlSqlServer, "Production.ProductCategory", connectionPropertiesSqlServer)

    println("Extracting Production.ProductSubcategory...")
    val dfProductSubcategory = spark.read.jdbc(jdbcUrlSqlServer, "Production.ProductSubcategory", connectionPropertiesSqlServer)

    println("Extracting Person.Address...")
    val dfAddress = spark.read.jdbc(jdbcUrlSqlServer, "Person.Address", connectionPropertiesSqlServer)

    println("Extracting Person.StateProvince...")
    val dfStateProvince = spark.read.jdbc(jdbcUrlSqlServer, "Person.StateProvince", connectionPropertiesSqlServer)

    println("Extracting Person.CountryRegion...")
    val dfCountryRegion = spark.read.jdbc(jdbcUrlSqlServer, "Person.CountryRegion", connectionPropertiesSqlServer)

    println("Extracting Sales.SalesOrderHeader...")
    val dfSalesOrderHeader = spark.read.jdbc(jdbcUrlSqlServer, "Sales.SalesOrderHeader", connectionPropertiesSqlServer)
      .withColumnRenamed("ModifiedDate", "HeaderModifiedDate_Raw")

    println("Extracting Sales.SalesOrderDetail...")
    val dfSalesOrderDetail = spark.read.jdbc(jdbcUrlSqlServer, "Sales.SalesOrderDetail", connectionPropertiesSqlServer)
      .withColumnRenamed("ModifiedDate", "DetailModifiedDate_Raw")

    println("\n--- All source tables extracted. ---")


    // --- PHASE 2: Transformation des Données ---
    println("\n--- Phase 2: Transformations des données ---")

    // --- 2.1. DimCustomer ---
    println("Creating DimCustomer...")
    val dfPersonForCustomer = dfPerson.withColumnRenamed("ModifiedDate", "Person_ModifiedDate")
                                       .withColumnRenamed("BusinessEntityID", "Person_BusinessEntityID")

    val dfDimCustomerTemp = dfCustomer.join(
      dfPersonForCustomer,
      dfCustomer("PersonID") === dfPersonForCustomer("Person_BusinessEntityID"),
      "left_outer"
    ).join(
      dfPersonPhone,
      dfCustomer("PersonID") === dfPersonPhone("BusinessEntityID"),
      "left_outer"
    )

    val windowSpecCustomer = Window.orderBy(col("CustomerID"))

    val dfDimCustomer = dfDimCustomerTemp.select(
      col("CustomerID"),
      col("PersonID"),
      col("StoreID"),
      col("TerritoryID"),
      col("AccountNumber"),
      concat_ws(" ", col("FirstName"), coalesce(col("MiddleName"), lit("")), col("LastName")).alias("Name"),
      col("EmailPromotion").cast(BooleanType).alias("EmailPromotion"),
      col("PhoneNumber").alias("PhoneNumber"),
      coalesce(dfCustomer("ModifiedDate"), col("Person_ModifiedDate")).alias("ModifiedDate")
    ).withColumn("CustomerKey", row_number().over(windowSpecCustomer))
     .dropDuplicates("CustomerKey")
     .select(
      col("CustomerKey").cast(IntegerType),
      col("CustomerID").cast(IntegerType),
      col("PersonID").cast(IntegerType),
      col("StoreID").cast(IntegerType),
      col("TerritoryID").cast(IntegerType),
      col("AccountNumber"),
      col("Name"),
      col("EmailPromotion"),
      col("PhoneNumber"),
      col("ModifiedDate").cast(DateType)
    )
    println("DimCustomer created.")


    // --- 2.2. DimProduct ---
    println("Creating DimProduct...")
    val dfProductCategoryRenamed = dfProductCategory.withColumnRenamed("Name", "ProductCategoryName")
                                                     .withColumnRenamed("ModifiedDate", "ProductCategoryModifiedDate")
    val dfProductSubcategoryRenamed = dfProductSubcategory.withColumnRenamed("Name", "ProductSubcategoryName")
                                                           .withColumnRenamed("ModifiedDate", "ProductSubcategoryModifiedDate")

    val dfDimProductTemp = dfProduct.join(
      dfProductSubcategoryRenamed,
      dfProduct("ProductSubcategoryID") === dfProductSubcategoryRenamed("ProductSubcategoryID"),
      "left_outer"
    ).join(
      dfProductCategoryRenamed,
      dfProductSubcategoryRenamed("ProductCategoryID") === dfProductCategoryRenamed("ProductCategoryID"),
      "left_outer"
    )

    val windowSpecProduct = Window.orderBy(col("ProductID"))

    val dfDimProduct = dfDimProductTemp.select(
      col("ProductID"),
      col("Name").alias("Name"),
      col("ProductNumber"),
      col("MakeFlag").cast(BooleanType),
      col("FinishedGoodsFlag").cast(BooleanType),
      col("Color"),
      col("SafetyStockLevel").cast(IntegerType),
      col("ReorderPoint").cast(IntegerType),
      col("StandardCost").cast(DecimalType(10,4)),
      col("ListPrice").cast(DecimalType(10,4)),
      col("Size"),
      col("SizeUnitMeasureCode"),
      col("Weight").cast(DecimalType(10,2)),
      col("WeightUnitMeasureCode"),
      col("DaysToManufacture").cast(IntegerType),
      col("ProductLine"),
      col("Class"),
      col("Style"),
      col("ProductCategoryName"),
      col("ProductSubcategoryName"),
      col("ModifiedDate").alias("ModifiedDate").cast(DateType)
    ).withColumn("ProductKey", row_number().over(windowSpecProduct))
     .dropDuplicates("ProductKey")
     .select(
      col("ProductKey"),
      col("ProductID"),
      col("Name"),
      col("ProductNumber"),
      col("MakeFlag"),
      col("FinishedGoodsFlag"),
      col("Color"),
      col("SafetyStockLevel"),
      col("ReorderPoint"),
      col("StandardCost"),
      col("ListPrice"),
      col("Size"),
      col("SizeUnitMeasureCode"),
      col("Weight"),
      col("WeightUnitMeasureCode"),
      col("DaysToManufacture"),
      col("ProductLine"),
      col("Class"),
      col("Style"),
      col("ProductCategoryName"),
      col("ProductSubcategoryName"),
      col("ModifiedDate")
    )
    println("DimProduct created.")


    // --- 2.3. DimGeography ---
    println("Creating DimGeography...")
    val dfStateProvinceRenamed = dfStateProvince.withColumnRenamed("Name", "StateProvinceName")
                                                 .withColumnRenamed("ModifiedDate", "StateModifiedDate")
    val dfCountryRegionRenamed = dfCountryRegion.withColumnRenamed("Name", "CountryRegionName")
                                                 .withColumnRenamed("ModifiedDate", "CountryModifiedDate")
                                                 .withColumnRenamed("CountryRegionCode", "CountryRegionCode_From_Country")

    val dfDimGeographyTemp = dfAddress.join(
      dfStateProvinceRenamed,
      dfAddress("StateProvinceID") === dfStateProvinceRenamed("StateProvinceID"),
      "left_outer"
    ).join(
      dfCountryRegionRenamed,
      dfStateProvinceRenamed("CountryRegionCode") === dfCountryRegionRenamed("CountryRegionCode_From_Country"),
      "left_outer"
    )

    val windowSpecGeography = Window.orderBy(col("AddressID"))

    val dfDimGeography = dfDimGeographyTemp.select(
      col("AddressID"),
      col("AddressLine1"),
      col("AddressLine2"),
      col("City"),
      col("StateProvinceCode"),
      col("StateProvinceName"),
      col("CountryRegionCode"),
      col("CountryRegionName"),
      col("PostalCode"),
      col("ModifiedDate").alias("ModifiedDate").cast(DateType)
    ).withColumn("GeographyKey", row_number().over(windowSpecGeography))
     .dropDuplicates("GeographyKey")
     .select(
      col("GeographyKey").cast(IntegerType),
      col("AddressID").cast(IntegerType),
      col("AddressLine1"),
      col("AddressLine2"),
      col("City"),
      col("StateProvinceCode"),
      col("StateProvinceName"),
      col("CountryRegionCode"),
      col("CountryRegionName"),
      col("PostalCode"),
      col("ModifiedDate")
    )
    println("DimGeography created.")


    // --- 2.4. DimDate ---
    println("Creating DimDate...")
    val dfDatesRaw = dfSalesOrderHeader.select(to_date(col("OrderDate")).alias("FullDateAlternateKey"))
      .union(dfSalesOrderHeader.select(to_date(col("DueDate")).alias("FullDateAlternateKey")))
      .union(dfSalesOrderHeader.select(to_date(col("ShipDate")).alias("FullDateAlternateKey")))
      .union(dfSalesOrderDetail.select(to_date(col("DetailModifiedDate_Raw")).alias("FullDateAlternateKey")))
      .distinct()
      .filter(col("FullDateAlternateKey").isNotNull)
      .orderBy(col("FullDateAlternateKey"))

    val dfDimDate = dfDatesRaw.select(
      col("FullDateAlternateKey"),
      dayofmonth(col("FullDateAlternateKey")).alias("DayNumberOfMonth"),
      date_format(col("FullDateAlternateKey"), "EEEE").alias("DayNameOfWeek"),
      dayofweek(col("FullDateAlternateKey")).alias("DayNumberOfWeek"),
      date_format(col("FullDateAlternateKey"), "D").alias("DayNumberOfYear"),
      weekofyear(col("FullDateAlternateKey")).alias("WeekNumberOfYear"),
      month(col("FullDateAlternateKey")).alias("MonthNumberOfYear"),
      date_format(col("FullDateAlternateKey"), "MMMM").alias("MonthName"),
      quarter(col("FullDateAlternateKey")).alias("QuarterNumberOfYear"),
      concat_ws(" ", lit("Q"), quarter(col("FullDateAlternateKey"))).alias("QuarterName"),
      year(col("FullDateAlternateKey")).alias("CalendarYear"),
      when(quarter(col("FullDateAlternateKey")).isin(1, 2), lit("H1"))
        .otherwise(lit("H2")).alias("CalendarSemester"),
      col("QuarterNumberOfYear").alias("FiscalQuarter"),
      col("CalendarYear").alias("FiscalYear"),
      col("CalendarSemester").alias("FiscalSemester"),
      col("FullDateAlternateKey").alias("ModifiedDate").cast(DateType)
    ).withColumn(
      "DateKey",
      (year(col("FullDateAlternateKey")) * 10000 +
       month(col("FullDateAlternateKey")) * 100 +
       dayofmonth(col("FullDateAlternateKey"))
      ).cast(IntegerType)
    ).dropDuplicates("DateKey")
     .select(
      col("DateKey").cast(IntegerType),
      col("FullDateAlternateKey").cast(DateType),
      col("DayNumberOfMonth").cast(IntegerType),
      col("DayNameOfWeek"),
      col("DayNumberOfWeek").cast(IntegerType),
      col("DayNumberOfYear").cast(IntegerType),
      col("WeekNumberOfYear").cast(IntegerType),
      col("MonthNumberOfYear").cast(IntegerType),
      col("MonthName"),
      col("QuarterNumberOfYear").cast(IntegerType),
      col("QuarterName"),
      col("CalendarYear").cast(IntegerType),
      col("CalendarSemester"),
      col("FiscalQuarter").cast(IntegerType),
      col("FiscalYear").cast(IntegerType),
      col("FiscalSemester"),
      col("ModifiedDate").cast(DateType)
    )
    println("DimDate created.")


    // --- 2.5. FactCustomSales ---
    println("Creating FactCustomSales...")

    val dfFactSalesTemp1 = dfSalesOrderHeader.join(
      dfSalesOrderDetail,
      dfSalesOrderHeader("SalesOrderID") === dfSalesOrderDetail("SalesOrderID"),
      "inner"
    )

    val dfFactSalesTemp2 = dfFactSalesTemp1.join(
      dfDimCustomer.select("CustomerKey", "CustomerID"), // Select only necessary columns
      dfFactSalesTemp1("CustomerID") === dfDimCustomer("CustomerID"),
      "inner"
    )

    val dfFactSalesTemp3 = dfFactSalesTemp2.join(
      dfDimProduct.select("ProductKey", "ProductID"), // Select only necessary columns
      dfFactSalesTemp2("ProductID") === dfDimProduct("ProductID"),
      "inner"
    )

    val dfFactSalesTemp4 = dfFactSalesTemp3.join(
      dfDimGeography.select("GeographyKey", "AddressID"), // Select only necessary columns
      dfFactSalesTemp3("ShipToAddressID") === dfDimGeography("AddressID"),
      "left_outer"
    )

    val dfFactSalesTemp5 = dfFactSalesTemp4.withColumn("OrderDateKey",
      (year(col("OrderDate")) * 10000 + month(col("OrderDate")) * 100 + dayofmonth(col("OrderDate"))).cast(IntegerType)
    ).join(
      dfDimDate.select("DateKey"), // Select only necessary columns
      col("OrderDateKey") === dfDimDate("DateKey"),
      "inner"
    )

    val dfFactCustomSales = dfFactSalesTemp5.select(
      col("CustomerKey").cast(IntegerType),
      col("ProductKey").cast(IntegerType),
      col("GeographyKey").cast(IntegerType),
      col("DateKey").alias("DateKey").cast(IntegerType),
      dfSalesOrderHeader("SalesOrderID").alias("SalesOrderID").cast(IntegerType),
      dfSalesOrderDetail("SalesOrderDetailID").alias("SalesOrderDetailID").cast(IntegerType),
      col("OrderQty").cast(IntegerType),
      col("UnitPrice").cast(DecimalType(10,4)),
      col("UnitPriceDiscount").cast(DecimalType(10,4)),
      col("LineTotal").cast(DecimalType(10,4)),
      col("OrderDate").cast(DateType),
      col("DueDate").cast(DateType),
      col("ShipDate").cast(DateType),
      col("OnlineOrderFlag").cast(BooleanType),
      col("SalesOrderNumber"),
      col("CarrierTrackingNumber"),
      col("HeaderModifiedDate_Raw").alias("ModifiedDate").cast(DateType)
    ).dropDuplicates("SalesOrderID", "SalesOrderDetailID")
    println("FactCustomSales created.")

    println("\n--- All dimensions and fact table transformations completed. ---")


    // --- PHASE 3: Chargement des données transformées dans PostgreSQL ---
    println("\n--- Phase 3: Chargement des données dans PostgreSQL Data Warehouse ---")

    // --- Écriture des dimensions ---
    println("Writing DimDate to PostgreSQL...")
    dfDimDate.write.mode(SaveMode.Append).jdbc(jdbcUrlPg, "public.DimDate", connectionPropertiesPg)
    println("DimDate loaded to PostgreSQL.")

    println("Writing DimCustomer to PostgreSQL...")
    dfDimCustomer.write.mode(SaveMode.Append).jdbc(jdbcUrlPg, "public.DimCustomer", connectionPropertiesPg)
    println("DimCustomer loaded to PostgreSQL.")

    println("Writing DimProduct to PostgreSQL...")
    dfDimProduct.write.mode(SaveMode.Append).jdbc(jdbcUrlPg, "public.DimProduct", connectionPropertiesPg)
    println("DimProduct loaded to PostgreSQL.")

    println("Writing DimGeography to PostgreSQL...")
    dfDimGeography.write.mode(SaveMode.Append).jdbc(jdbcUrlPg, "public.DimGeography", connectionPropertiesPg)
    println("DimGeography loaded to PostgreSQL.")

    // --- Écriture de la table de faits ---
    println("Writing FactCustomSales to PostgreSQL...")
    // Attention: Pour les tables de faits volumineuses, le mode "overwrite" peut être lent car il supprime et recrée la table.
    // Pour des mises à jour incrémentales, il faudrait une logique plus complexe (upsert, partitionnement).
    dfFactCustomSales.write.mode(SaveMode.Append).jdbc(jdbcUrlPg, "public.FactCustomSales", connectionPropertiesPg)
    println("FactCustomSales loaded to PostgreSQL.")

    println("\n--- All data loaded into PostgreSQL Data Warehouse! ---")


    // --- PHASE 4: Agrégations et écriture en CSV (Facultatif, pour rapport) ---
    // Ces agrégations sont exécutées après le chargement pour des raisons de clarté
    // mais pourraient être faites avant si les rapports sont les seuls consommateurs.
    println("\n--- Phase 4: Performing Aggregations and saving to CSV ---")

    // 1. Chiffre d'affaires par catégorie de produit
    println("Calculating Total Revenue by Product Category...")
    val dfRevenueByCategory = dfFactCustomSales.join(
      dfDimProduct.select("ProductKey", "ProductCategoryName"),
      Seq("ProductKey"),
      "inner"
    ).groupBy("ProductCategoryName").agg(
      sum("LineTotal").alias("TotalRevenue")
    ).orderBy(col("TotalRevenue").desc)
    dfRevenueByCategory.write.mode("overwrite").option("header", "true").csv("/app/output/aggregations/revenue_by_category")
    println("Total Revenue by Product Category saved to CSV.")


    // 2. Chiffre d'affaires par ville
    println("Calculating Total Revenue by City...")
    val dfRevenueByCity = dfFactCustomSales.join(
      dfDimGeography.select("GeographyKey", "City"),
      Seq("GeographyKey"),
      "inner"
    ).groupBy("City").agg(
      sum("LineTotal").alias("TotalRevenue")
    ).orderBy(col("TotalRevenue").desc)
    dfRevenueByCity.write.mode("overwrite").option("header", "true").csv("/app/output/aggregations/revenue_by_city")
    println("Total Revenue by City saved to CSV.")


    // 3. Chiffre d'affaires par année
    println("Calculating Total Revenue by Year...")
    val dfRevenueByYear = dfFactCustomSales.join(
      dfDimDate.select("DateKey", "CalendarYear"),
      dfFactCustomSales("DateKey") === dfDimDate("DateKey"),
      "inner"
    ).groupBy("CalendarYear").agg(
      sum("LineTotal").alias("TotalRevenue")
    ).orderBy(col("CalendarYear"))
    dfRevenueByYear.write.mode("overwrite").option("header", "true").csv("/app/output/aggregations/revenue_by_year")
    println("Total Revenue by Year saved to CSV.")


    println("\n--- Full ETL pipeline completed successfully! ---")

    // Arrêter la session Spark à la fin du script
    spark.stop()
  }
}

File: etl/etl_loading.py
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, concat_ws, lit, to_date, year, month, dayofmonth, dayofweek, weekofyear, quarter, date_format, lpad, when, coalesce, sum
from pyspark.sql.types import IntegerType, DateType, BooleanType, DecimalType,  ShortType, StringType
from pyspark.sql.window import Window
from pyspark.sql.functions import row_number


# 1. Initialisation de la session Spark
spark = SparkSession.builder \
    .appName("AdventureWorks ETL") \
    .config("spark.jars", "/app/lib/mssql-jdbc-12.8.1.jre11.jar,/app/lib/postgresql-42.7.3.jar") \
    .getOrCreate()


# 2. Paramètres JDBC
jdbc_url = "jdbc:sqlserver://sqlserver:1433;databaseName=AdventureWorks;encrypt=true;trustServerCertificate=true"
connection_properties = {
    "user": "SA",
    "password": "password123?",
    "driver": "com.microsoft.sqlserver.jdbc.SQLServerDriver"
}

print("Starting data extraction from AdventureWorks...")

# --- 3. Extraction des tables sources ---
print("Extracting Sales.Customer...")
df_customer = spark.read.jdbc(url=jdbc_url, table="Sales.Customer", properties=connection_properties)
print("Sales.Customer extracted.")

print("Extracting Person.Person...")
df_person = spark.read.jdbc(url=jdbc_url, table="Person.Person", properties=connection_properties)
print("Person.Person extracted.")

print("Extracting Person.EmailAddress...")
df_email_address = spark.read.jdbc(url=jdbc_url, table="Person.EmailAddress", properties=connection_properties)
print("Person.EmailAddress extracted.")


print("Extracting Sales.SalesOrderHeader...")
df_sales_order_header = spark.read.jdbc(url=jdbc_url, table="Sales.SalesOrderHeader", properties=connection_properties)
print("Sales.SalesOrderHeader extracted.")

print("Extracting Sales.SalesOrderDetail...")
df_sales_order_detail = spark.read.jdbc(url=jdbc_url, table="Sales.SalesOrderDetail", properties=connection_properties)
print("Sales.SalesOrderDetail extracted.")

print("Extracting Production.Product...")
df_product = spark.read.jdbc(url=jdbc_url, table="Production.Product", properties=connection_properties)
print("Production.Product extracted.")

print("Extracting Production.ProductCategory...")
df_product_category = spark.read.jdbc(url=jdbc_url, table="Production.ProductCategory", properties=connection_properties)
print("Production.ProductCategory extracted.")

print("Extracting Production.ProductSubcategory...")
df_product_subcategory = spark.read.jdbc(url=jdbc_url, table="Production.ProductSubcategory", properties=connection_properties)
print("Production.ProductSubcategory extracted.")

print("Extracting Person.Address...")
df_address = spark.read.jdbc(url=jdbc_url, table="Person.Address", properties=connection_properties)
print("Person.Address extracted.")

print("Extracting Person.StateProvince...")
df_state_province = spark.read.jdbc(url=jdbc_url, table="Person.StateProvince", properties=connection_properties)
print("Person.StateProvince extracted.")

print("Extracting Person.CountryRegion...")
df_country_region = spark.read.jdbc(url=jdbc_url, table="Person.CountryRegion", properties=connection_properties)
print("Person.CountryRegion extracted.")


print("\n--- 4. Transformations et Création des Dimensions et de la table de Faits ---")

# --- DimCustomer ---
print("Creating DimCustomer...")
window_spec_customer = Window.orderBy(col("CustomerID"))
# Création d'un DataFrame temporaire qui inclut CustomerID pour la jointure avec la table de faits
df_dim_customer_enriched = df_customer.alias("c").join(
    df_person.alias("p"),
    col("c.PersonID") == col("p.BusinessEntityID"),
    "left"
).join(
    df_email_address.alias("ea"),
    col("p.BusinessEntityID") == col("ea.BusinessEntityID"),
    "left"
).select(
    col("c.CustomerID"), # Garder CustomerID pour la jointure avec la table de faits
    row_number().over(window_spec_customer).cast(IntegerType()).alias("CustomerKey"),
    col("p.FirstName"),
    col("p.LastName"),
    col("ea.EmailAddress"),
    when(col("c.StoreID").isNotNull(), lit("Store")).otherwise(lit("Individual")).alias("CustomerType")
)

# DataFrame final pour DimCustomer (selon le DDL)
df_dim_customer = df_dim_customer_enriched.select(
    col("CustomerKey"),
    col("FirstName").cast(StringType()).alias("FirstName"),
    col("LastName").cast(StringType()).alias("LastName"),
    col("EmailAddress").cast(StringType()).alias("EmailAddress"),
    col("CustomerType").cast(StringType()).alias("CustomerType")
)
print("DimCustomer created.")
df_dim_customer.printSchema()
df_dim_customer.show(5)


# --- DimProduct ---
print("Creating DimProduct...")
window_spec_product = Window.orderBy(col("ProductID"))
# Création d'un DataFrame temporaire qui inclut ProductID pour la jointure avec la table de faits
df_dim_product_enriched = df_product.alias("p").join(
    df_product_subcategory.alias("psc"),
    col("p.ProductSubcategoryID") == col("psc.ProductSubcategoryID"),
    "left"
).join(
    df_product_category.alias("pc"),
    col("psc.ProductCategoryID") == col("pc.ProductCategoryID"),
    "left"
).select(
    col("p.ProductID"), # Garder ProductID pour la jointure avec la table de faits
    row_number().over(window_spec_product).cast(IntegerType()).alias("ProductKey"),
    col("p.Name").alias("ProductName"),
    col("p.ProductNumber"),
    col("pc.Name").alias("ProductCategory"),
    col("psc.Name").alias("ProductSubcategory")
)

# DataFrame final pour DimProduct (selon le DDL)
df_dim_product = df_dim_product_enriched.select(
    col("ProductKey"),
    col("ProductName").cast(StringType()).alias("ProductName"),
    col("ProductNumber").cast(StringType()).alias("ProductNumber"),
    col("ProductCategory").cast(StringType()).alias("ProductCategory"),
    col("ProductSubcategory").cast(StringType()).alias("ProductSubcategory")
)
print("DimProduct created.")
df_dim_product.printSchema()
df_dim_product.show(5)


# --- DimGeography ---
print("Creating DimGeography...")
window_spec_geography = Window.orderBy(col("AddressID"))
# Création d'un DataFrame temporaire qui inclut AddressID pour la jointure avec la table de faits
df_dim_geography_enriched = df_address.alias("a").join(
    df_state_province.alias("sp"),
    col("a.StateProvinceID") == col("sp.StateProvinceID"),
    "left"
).join(
    df_country_region.alias("cr"),
    col("sp.CountryRegionCode") == col("cr.CountryRegionCode"),
    "left"
).select(
    col("a.AddressID"), # Garder AddressID pour la jointure avec la table de faits
    row_number().over(window_spec_geography).cast(IntegerType()).alias("GeographyKey"),
    col("a.City"),
    col("sp.Name").alias("StateProvince"),
    col("cr.Name").alias("CountryRegion"),
    col("a.PostalCode")
)

# DataFrame final pour DimGeography (selon le DDL)
df_dim_geography = df_dim_geography_enriched.select(
    col("GeographyKey"),
    col("City").cast(StringType()).alias("City"),
    col("StateProvince").cast(StringType()).alias("StateProvince"),
    col("CountryRegion").cast(StringType()).alias("CountryRegion"),
    col("PostalCode").cast(StringType()).alias("PostalCode")
)
print("DimGeography created.")
df_dim_geography.printSchema()
df_dim_geography.show(5)


# --- DimDate ---
print("Creating DimDate...")
# Créer un DataFrame de dates unique à partir de OrderDate
df_dates = df_sales_order_header.select(col("OrderDate").alias("FullDate")).distinct()
window_spec_date = Window.orderBy(col("FullDate"))
# Création d'un DataFrame temporaire qui inclut FullDate pour la jointure avec la table de faits
df_dim_date_enriched = df_dates.select(
    col("FullDate"), # Garder FullDate pour la jointure avec la table de faits
    (year(col("FullDate")) * 10000 + month(col("FullDate")) * 100 + dayofmonth(col("FullDate"))).cast(IntegerType()).alias("DateKey"),
    year(col("FullDate")).cast(IntegerType()).alias("Year"),
    month(col("FullDate")).cast(IntegerType()).alias("Month"),
    date_format(col("FullDate"), "MMMM").alias("MonthName"),
    quarter(col("FullDate")).cast(IntegerType()).alias("Quarter")
)

# DataFrame final pour DimDate (selon le DDL)
df_dim_date = df_dim_date_enriched.select(
    col("DateKey"),
    col("FullDate"),
    col("Year"),
    col("Month"),
    col("MonthName").cast(StringType()).alias("MonthName"),
    col("Quarter")
)
print("DimDate created.")
df_dim_date.printSchema()
df_dim_date.show(5)


# --- FactCustomSales ---
print("Creating FactCustomSales...")
df_fact_custom_sales = df_sales_order_detail.alias("sod").join(
    df_sales_order_header.alias("soh"),
    col("sod.SalesOrderID") == col("soh.SalesOrderID"),
    "inner"
).join(
    # Jointure avec la dimension client enrichie pour obtenir CustomerKey
    df_dim_customer_enriched.alias("dc"), # Alias pour la dimension enrichie
    col("soh.CustomerID") == col("dc.CustomerID"), # Utiliser l'alias dc.CustomerID
    "inner"
).join(
    # Jointure avec la dimension produit enrichie pour obtenir ProductKey
    df_dim_product_enriched.alias("dp"), # Alias pour la dimension enrichie
    col("sod.ProductID") == col("dp.ProductID"), # Utiliser l'alias dp.ProductID
    "inner"
).join(
    # Jointure avec la dimension géographie enrichie pour obtenir GeographyKey
    df_dim_geography_enriched.alias("dg"), # Alias pour la dimension enrichie
    col("soh.BillToAddressID") == col("dg.AddressID"), # Utiliser l'alias dg.AddressID
    "inner"
).join(
    # Jointure avec la dimension date enrichie pour obtenir DateKey
    df_dim_date_enriched.alias("dd"), # Alias pour la dimension enrichie
    to_date(col("soh.OrderDate")) == col("dd.FullDate"), # Utiliser l'alias dd.FullDate
    "inner"
).select(
    col("dc.CustomerKey"), # Sélectionner les clés des dimensions enrichies avec leurs alias
    col("dp.ProductKey"),
    col("dg.GeographyKey"),
    col("dd.DateKey"),
    col("sod.OrderQty").cast(ShortType()).alias("OrderQuantity"),
    col("sod.LineTotal").cast(DecimalType(19, 4)).alias("SalesAmount"),
    col("sod.UnitPrice").cast(DecimalType(19, 4)).alias("UnitPrice")
)

# Appliquer le schéma exact du DDL pour FactCustomSales
df_fact_custom_sales = df_fact_custom_sales.select(
    col("CustomerKey"),
    col("ProductKey"),
    col("GeographyKey"),
    col("DateKey"),
    col("OrderQuantity"),
    col("SalesAmount"),
    col("UnitPrice")
)
print("FactCustomSales created.")
df_fact_custom_sales.printSchema()
df_fact_custom_sales.show(5)


# --- 5. Chargement des Données Transformées dans PostgreSQL ---
print("\n--- Starting data loading into PostgreSQL ---")

# Paramètres JDBC pour PostgreSQL
#jdbc_url_pg = "jdbc:postgresql://postgresql:5432/adventure_warehouse"
jdbc_url_pg = "jdbc:postgresql://host.docker.internal:5432/adventure_warehouse"
connection_properties_pg = {
    "user": "postgres",
    "password": "pass",
    "driver": "org.postgresql.Driver",
    "stringtype": "unspecified"
}






# Charger et écrire FactCustomSales
print("Loading FactCustomSales to PostgreSQL...")
df_fact_custom_sales.write.jdbc(url=jdbc_url_pg, table='public."FactCustomSales"', mode="overwrite", properties=connection_properties_pg)
print("FactCustomSales loaded to PostgreSQL.")

# Charger et écrire DimDate
print("Loading DimDate to PostgreSQL...")
df_dim_date.write.jdbc(url=jdbc_url_pg, table='public."DimDate"', mode="overwrite", properties=connection_properties_pg)
print("DimDate loaded to PostgreSQL.")

# Charger et écrire DimCustomer
print("Loading DimCustomer to PostgreSQL...")
df_dim_customer.write.jdbc(url=jdbc_url_pg, table='public."DimCustomer"', mode="overwrite", properties=connection_properties_pg)
print("DimCustomer loaded to PostgreSQL.")

# Charger et écrire DimProduct
print("Loading DimProduct to PostgreSQL...")
df_dim_product.write.jdbc(url=jdbc_url_pg, table='public."DimProduct"', mode="overwrite", properties=connection_properties_pg)
print("DimProduct loaded to PostgreSQL.")

# Charger et écrire DimGeography
print("Loading DimGeography to PostgreSQL...")
df_dim_geography.write.jdbc(url=jdbc_url_pg, table='public."DimGeography"', mode="overwrite", properties=connection_properties_pg)
print("DimGeography loaded to PostgreSQL.")

print("\n--- All data loaded into PostgreSQL. ---")

# Arrêter la session Spark
spark.stop()
print("Spark session stopped.")